src/plugins/es_ui_shared/static/forms/hook_form_lib/components/use_field.tsx

```diff
@@ -102,7 +102,7 @@ function UseFieldComp<T = unknown>(props: Props<T>) {
     );
   }
 
-  return componentToRender({ field, ...propsToForward });
+  return <ComponentToRender {...{ field, ...propsToForward }} />;
```

Comment:
@sebelga I haven't yet added a test to encompass this change, but you can read about the errant behavior in https://github.com/elastic/kibana/pull/76831/commits/9d5ce1c5a594e845067f486c72ae51ce1c91b592. My first thought was simply to assert that `React.createElement` is called when a custom component is used, but I'd appreciate your thoughts!

Reply:
I would prefer to test the different use cases instead of the implementation. So I would have a dummy form with 3 types of fields (or more if there are more use cases) and make sure they render correctly (are present in the DOM)

* no component passed --> renders an input in the dom
* component passed --> renders it
* memoized component passed --> renders it

Reply:
Do you mind explaining what you mean by "_Cannot switch between UseField components_"?

Reply:
@sebelga in this PR we're swapping between two different field components based on which rule type the user has chosen. [link](https://github.com/elastic/kibana/pull/76831/files#diff-4fcc20d44b08313797368858a805cfe5R233-R258). This causes react warnings if the components are not invoked with React.createElement.

Reply:
@sebelga tests added here: a9a6703

Reply:
Cool thanks for the tests @rylnd !

For the swap of components, we took the habit of adding a `key` to the `UseField` being swapped. To make sure the component unmounts and re-mounts.